(#220) Write, build and load a business logic module for Logos (#227)
status-im/ift-docs: docs/core/journeys/build-run-a-logos-core-module.md

---
title: Build and run a Logos core module
doc_type: procedure
product: core
topics: core
steps_layout: sectioned
authors: iurimatias, Khushboo-dev-cpp, cheny0
owner: logos
doc_version: 1
slug: build-run-a-logos-core-module
---

# Build and run a Logos core module

#### Scaffold, build, package, and test a core module on the Logos platform.

> [!NOTE]
>
> - **Permissions**: No special permissions required.
> - **Product**: Logos Basecamp

The Logos platform is a modular application framework built on Qt 6. Applications are composed of dynamically loaded modules (Qt plugins) that provide features and functionality.

Logos core modules are non-UI modules that provide backend functionality. Core modules run in isolated `logos_host` processes and communicate via Qt Remote Objects.

> [!NOTE]
>

> > For other module types, check out [Wrap a C Library as a Logos core module](./wrap-a-c-library-as-a-logos-core-module.md), [Build a QML UI for your logos module](./build-a-qml-ui-for-your-logos-module.md) and [Build a Logos C++ UI module](./build-a-logos-cpp-ui-module.md).

Reviewer: wrong link

```suggestion
> For other module types, check out [Wrap a C Library as a Logos core module](./wrap-a-c-library-as-a-logos-module.md), [Build a QML UI for your logos module](./build-a-qml-ui-for-your-logos-module.md) and [Build a Logos C++ UI module](./build-a-logos-cpp-ui-module.md).
```

Reviewer 2: Actually, in the PR for this tutorial, it's called a logos core module: https://github.com/logos-co/logos-docs/pull/239. I'll remove the file in this PR.

## Step 2: Adapt the template for your module

The template generates files with placeholder names like `my_module` and `doSomething`. Replace these in every generated file to match your module's name and methods.

1. Edit `metadata.json` and set `name`, `version`, `description`, and `main` to match your module.
   - `name` must be a valid C identifier; it is used in filenames, method calls, and module loading.
   - `main` must match the plugin filename without the extension (for example, `my_module_plugin` resolves to `my_module_plugin.so` or `.dylib`).

1. Edit `CMakeLists.txt` and update the `project()` name and the `NAME` and `SOURCES` values inside the `logos_module()` call.
   - The `CMakeLists.txt` uses the `logos_module()` macro to handle Qt plugin setup.
   - `NAME` must match the `name` field in `metadata.json`. A mismatch causes the build to succeed but the install phase to fail.

1. Edit `flake.nix` and update the `description` field.
   - The generated `flake.nix` uses an unpinned `logos-module-builder` URL. For reproducible builds, pin it to `tutorial-v1`.

1. Rename the source files in `src/` to match your module name.

>    ```bash
>    mv src/my_module_interface.h    src/<module_name>_interface.h
>    mv src/my_module_plugin.h       src/<module_name>_plugin.h
>    mv src/my_module_plugin.cpp     src/<module_name>_plugin.cpp
>    ```

Reviewer: seems like there is different placeholder 

```suggestion
   ```bash
   mv src/minimal_interface.h    src/<module_name>_interface.h
   mv src/minimal_plugin.h       src/<module_name>_plugin.h
   mv src/minimal_plugin.cpp     src/<module_name>_plugin.cpp
   ```
```

   ```

## Step 5: Package the module

Before you can run your module with `logoscore` or install it into `logos-basecamp`, you need to package the build output into an `.lgx` package and install it into a `modules/` directory. Check out the [LGX package format and bundling reference](./lgx-package-format-and-bundling-reference.md) for more details on the format and bundling options.

> [!NOTE]
>
> The `manifest.json` is auto-generated from your module's `metadata.json` by the bundler. It maps each variant to its main entry point.

There are two ways to create `.lgx` packages:

- Use the built-in Nix derivation that comes with `logos-module-builder` (preferred). 
- Use the `nix bundle` command directly.

### Use the Nix derivation

When your module uses `logos-module-builder`, LGX package outputs are automatically available as part of your flake (the builder includes `nix-bundle-lgx` internally).

1. Bundle the module into an LGX package that uses `/nix/store` references for local development.

   ```bash
   nix build .#lgx
   ```

   - Use `#lgx-portable` for a self-contained, all dependencies bundled package: `nix build .#lgx-portable`.

> 1. Check the `result/` directory and confirm the `<module-name>-<version>.lgx` file is present.

Reviewer: nix produces different naming, so this is correct 

```suggestion
1. Check the `result/` directory and confirm the `logos-<name>-module-lib.lgx` file is present.
```

### Use the `nix bundle` command

The `nix bundle` command is useful if your module does not use `logos-module-builder`, or if you need the `dual` bundling mode (both `dev` and `portable` in a single `.lgx` file) which is only available via the `nix bundle` command.

1. Bundle the module into an LGX package using the `nix bundle` command.

   ```bash
   nix bundle --bundler github:logos-co/nix-bundle-lgx/tutorial-v1 .#lib
   ```

   - Use `#portable` for a self-contained package with no `/nix/store` references: `nix bundle --bundler github:logos-co/nix-bundle-lgx/tutorial-v1#portable .#lib`.
   - Use `#dual` to produce both dev and portable variants in a single `.lgx` file: `nix bundle --bundler github:logos-co/nix-bundle-lgx/tutorial-v1#dual .#lib`.

> 1. Check the current directory and confirm the `<module-name>-<version>.lgx` file is present.

Reviewer: similar to https://github.com/logos-co/logos-docs/pull/227/changes#r3300272638
```suggestion
1. Check the current directory and confirm the `logos-<name>-module-lib.lgx` file is present.
```

   ```

1. Stop the daemon when finished.

   ```bash
   ./logos/bin/logoscore stop
   ```

> [!TIP]
>
> Check out [Logos CLI Reference](./logos-cli-reference.md) for more details on available commands and options.

### Run with `logos-basecamp`

Logos Basecamp is a desktop application that provides a graphical interface for managing and running modules. Core modules run as background services in `logos-basecamp`. Other UI modules can call them through `LogosAPI` or the `logos.callModule()` bridge once they are installed.

> [!IMPORTANT]
>
> The LGX variant type must match the basecamp build type. Dev builds of basecamp expect dev LGX variants (for example, `darwin-arm64-dev`), and portable builds expect portable variants (for example, `darwin-arm64`). Check out the [LGX package format and bundling reference](./lgx-package-format-and-bundling-reference.md) for more details.

1. Build the development version of `logos-basecamp`.

   ```bash
   nix build 'github:logos-co/logos-basecamp/tutorial-v1#app' --out-link ./logos-basecamp
   ```

1. Launch `logos-basecamp` once to create its data directory and preinstall bundled modules, then close it. 

   ```bash

>    ./basecamp-result/bin/logos-basecamp

Reviewer: this will be correct binary given prev step

```suggestion
    ./logos-basecamp/bin/logos-basecamp
```

   ```

>    - To find the data directory, check the log output for `plugins directory` or look for the directory containing `modules/` and `plugins/` subdirectories at `~/Library/Application Support/Logos/` (macOS) or `~/.local/share/Logos/` (Linux).

Reviewer: this is just seems to be wrong

```suggestion
   - To find the data directory, check the log output for `plugins directory` or look for the directory containing `modules/` and `plugins/` subdirectories at `~/Library/Application Support/Logos/` (macOS) or `.local/share/Logos/LogosBasecampDev` (Linux).
```

   ```bash
   cd <module-name>
   ./logos/bin/logoscore load-module <module-name>
   ./logos/bin/logoscore call <module-name> <method> <args>
   ```

1. Stop the daemon when finished.

   ```bash
   ./logos/bin/logoscore stop
   ```

> [!TIP]
>
> Check out [Logos CLI Reference](./logos-cli-reference.md) for more details on available commands and options.

### Run with `logos-basecamp`

Logos Basecamp is a desktop application that provides a graphical interface for managing and running modules. Core modules run as background services in `logos-basecamp`. Other UI modules can call them through `LogosAPI` or the `logos.callModule()` bridge once they are installed.

> [!IMPORTANT]
>
> The LGX variant type must match the basecamp build type. Dev builds of basecamp expect dev LGX variants (for example, `darwin-arm64-dev`), and portable builds expect portable variants (for example, `darwin-arm64`). Check out the [LGX package format and bundling reference](./lgx-package-format-and-bundling-reference.md) for more details.

1. Build the development version of `logos-basecamp`.

   ```bash

>    nix build 'github:logos-co/logos-basecamp/tutorial-v1#app' --out-link ./logos-basecamp

Reviewer 3: shouldn't it use the latest release of logos-basecamp?

Reviewer 3: cc @Khushboo-dev-cpp

Reviewer 4: it should. unless this doc is linked to tutorial-V1

Author: @fryorcraken @Khushboo-dev-cpp, the [original tutorial](https://github.com/logos-co/logos-tutorial/blob/tutorial-v1/logos-developer-guide.md#61-building-logos-basecamp) uses the `tutorial-v1` tag. Should I change it?

Reviewer 3: Yes, let's make the doc work for `tutorial-v1` as per the original tutorial, and `tutorial-v1` tag on the logos-tutorial repo
